sign.go

```diff
@@ -0,0 +1,101 @@
+package floodsub
+
+import (
+	"fmt"
+
+	pb "github.com/libp2p/go-floodsub/pb"
+
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+const SignPrefix = "libp2p-pubsub:"
+
+func verifyMessageSignature(m *pb.Message) error {
+	pubk, err := messagePubKey(m)
+	if err != nil {
+		return err
+	}
+
+	xm := pb.Message{
+		Data:     m.Data,
+		TopicIDs: m.TopicIDs,
+		From:     m.From,
+		Seqno:    m.Seqno,
+	}
+	bytes, err := xm.Marshal()
+	if err != nil {
+		return err
+	}
+
+	bytes = withSignPrefix(bytes)
+
+	valid, err := pubk.Verify(bytes, m.Signature)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return fmt.Errorf("invalid signature")
+	}
+
+	return nil
+}
+
+func messagePubKey(m *pb.Message) (crypto.PubKey, error) {
+	var pubk crypto.PubKey
+
+	pid, err := peer.IDFromBytes(m.From)
+	if err != nil {
+		return nil, err
+	}
+
+	if m.Key == nil {
+		// no attached key, it must be extractable from the source ID
+		pubk, err = pid.ExtractPublicKey()
+		if err != nil {
+			return nil, fmt.Errorf("cannot extract signing key: %s", err.Error())
+		}
+	} else {
+		pubk, err = crypto.UnmarshalPublicKey(m.Key)
+		if err != nil {
+			return nil, fmt.Errorf("cannot unmarshal signing key: %s", err.Error())
+		}
+
+		// verify that the source ID matches the attached key
+		if !pid.MatchesPublicKey(pubk) {
+			return nil, fmt.Errorf("bad signing key; source ID %s doesn't match key", pid)
+		}
+	}
+
+	return pubk, nil
+}
+
+func signMessage(key crypto.PrivKey, m *pb.Message) error {
+	bytes, err := m.Marshal()
+	if err != nil {
+		return err
+	}
+
+	bytes = withSignPrefix(bytes)
+
+	sig, err := key.Sign(bytes)
+	if err != nil {
+		return err
+	}
+
+	m.Signature = sig
+	switch key.Type() {
+	case crypto.RSA:
```

Comment:
Ah. Actually, the "correct" way to do this is to:

1. Generate a peer ID.
2. Try to extract the key from the peer ID (`ExtractPublicKey`).

We may (will) have other key types that don't "embed".

Reply:
It's also a lot more slower. Are we planning on adding other non-embeddable keys in the future?

Reply:
maybe that overhead is negligible in the grand scale of things though.

Reply:
We could add some "Embeds(key) bool` helper somewhere.

>  Are we planning on adding other non-embeddable keys in the future?

I don't think sekp256 keys embed and there's a PR in progress to add more general ECDSA keys that definitely won't embed.

Reply:
Ok, but let's define a helper function in go-libp2p-peer.

Reply:
for now I added an extraction test.